{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Technology-specific guidance\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Harpy is designed to support all spatial technologies at single-cell resolution, with a focus on spatial proteomics and targeted/untargeted transcriptomics.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Targeted transcriptomics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Readers\n",
    "- Xenium and Merscope: {py:func}`harpy.io.xenium`, {py:func}`harpy.io.merscope`\n",
    "- Transcripts only:\n",
    "  - Merscope: {py:func}`harpy.io.read_merscope_transcripts`\n",
    "  - Xenium and CosMx: {py:func}`harpy.io.read_transcripts`\n",
    "  - Resolve: {py:func}`harpy.io.read_resolve_transcripts`\n",
    "\n",
    "### Example datasets\n",
    "- Xenium: {py:func}`harpy.datasets.xenium_human_lung_cancer`, {py:func}`harpy.datasets.xenium_human_ovarian_cancer`\n",
    "- Merscope: {py:func}`harpy.datasets.merscope_mouse_liver`\n",
    "- Resolve: {py:func}`harpy.datasets.resolve_example`\n",
    "\n",
    "### Segmentation\n",
    "We recommend [Cellpose](https://github.com/MouseLand/cellpose) via {py:func}`harpy.im.segment` with {py:func}`harpy.im.cellpose_callable`. See the {py:func}`harpy.im.segment` docstring for an example.\n",
    "\n",
    "For multi-modal segmentation, use {py:func}`harpy.im.merge_labels_layers` or {py:func}`harpy.im.merge_labels_layers_nuclei` (tutorial coming soon).\n",
    "\n",
    "### Create the AnnData table\n",
    "Use {py:func}`harpy.tb.allocate` to create a count matrix from a segmentation mask and detected transcripts.\n",
    "\n",
    "...and much more. See the tutorials for more information.\n",
    "\n",
    "### Tutorials\n",
    "- Xenium: [Tutorial Human Ovarian Cancer, Xenium 10x Genomics](../advanced/Harpy_transcriptomics_xenium.ipynb)\n",
    "- Resolve: [Tutorial Mouse Liver, Resolve Molecular Cartography](../advanced/Harpy_transcriptomics.ipynb)\n",
    "- Merscope: coming soon\n",
    "- CosMx: coming soon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Untargeted transcriptomics\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Readers\n",
    "- Visium and VisiumHD: {py:func}`harpy.io.visium`, {py:func}`harpy.io.visiumHD`\n",
    "- Transcripts only:\n",
    "  - BGI Stereo-seq: {py:func}`harpy.io.read_stereoseq_transcripts`\n",
    "\n",
    "### Example datasets\n",
    "- VisiumHD: {py:func}`harpy.datasets.visium_hd_example` and {py:func}`harpy.datasets.visium_hd_example_custom_binning`.\n",
    "\n",
    "### Segmentation\n",
    "For custom segmentation on H&E, we recommend [Cellpose](https://github.com/MouseLand/cellpose) (version>4.0) via {py:func}`harpy.im.segment` with {py:func}`harpy.im.cellpose_callable`.\n",
    "\n",
    "### Custom binning\n",
    "For binning using a custom generated segmentation mask or raster we provide {py:func}`harpy.tb.bin_counts`.\n",
    "\n",
    "For creation of custom rasters (tesselation), we provide the utility function {py:func}`harpy.im.add_grid_labels_layer`.\n",
    "\n",
    "### Tutorials\n",
    "\n",
    "Coming soon."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spatial proteomics\n",
    "\n",
    "### Readers\n",
    "- MACSima: {py:func}`harpy.io.macsima`.\n",
    "\n",
    "### Example datasets\n",
    "- MACSima: {py:func}`harpy.datasets.macsima_colorectal_carcinoma`, {py:func}`harpy.datasets.macsima_tonsil` and {py:func}`harpy.datasets.macsima_example`.\n",
    "- CODEX: {py:func}`harpy.datasets.codex_example`.\n",
    "- MIBI-TOF: {py:func}`harpy.datasets.mibi_example`.\n",
    "- Vectra Polaris: {py:func}`harpy.datasets.vectra_example`.\n",
    "\n",
    "### Segmentation\n",
    "For segmentation we recommend either [Cellpose](https://github.com/MouseLand/cellpose) (version>4.0) or [InstanSeg](https://github.com/instanseg/instanseg), via {py:func}`harpy.im.segment` with {py:func}`harpy.im.cellpose_callable`, respectively {py:func}`harpy.im.instanseg_callable`,\n",
    "\n",
    "When Cellpose is used for segmentation, multiple segmentation masks, obtained through segmentation on various markers can be merged using {py:func}`harpy.im.merge_labels_layers` and {py:func}`harpy.im.merge_labels_layers_nuclei`.\n",
    "\n",
    "### Creation of the AnnData table.\n",
    "\n",
    "For creation of an AnnData table that contains mean intensity for every instance in the segmentation mask for every marker, we provide {py:func}`harpy.tb.allocate_intensity`.\n",
    "\n",
    "### Custom embedding\n",
    "\n",
    "We provide {py:func}`harpy.tb.featurize` to generate cell embeddings using a user-provided computer vision model.\n",
    "\n",
    "### And much more, see the tutorials\n",
    "\n",
    "### Tutorials\n",
    "- MACSima: [Segmentation, feature extraction and cell clustering](../general/Harpy_feature_calculation.ipynb); [Computational benchmark](../general/benchmark.ipynb).\n",
    "- Vectra Polaris: [Segmentation with InstanSeg](../general/Harpy_instanseg.ipynb).\n",
    "- CODEX: [Creating cell embeddings](../general/Featurize_with_kronos.ipynb); [generating single cell representations - integration with PyTorch](../general/generate_single_cell_representations.ipynb).\n",
    "- MIBI-TOF: [Pixel and cell clustering with FlowSOM](../general/FlowSOM_for_pixel_and_cell_clustering.ipynb).\n",
    "\n",
    "For API details, see the [API reference](../../api).\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv_harpy_1_12_25",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
